src/datatype.c

```diff
@@ -140,3 +159,798 @@ pg_tle_create_shell_type_if_not_exists(PG_FUNCTION_ARGS)
+ *
+ */
+PG_FUNCTION_INFO_V1(pg_tle_create_base_type_if_not_exists);
+Datum
+pg_tle_create_base_type_if_not_exists(PG_FUNCTION_ARGS)
+{
+	Oid			typeNamespace = PG_GETARG_OID(0);
+	char	   *typeName = NameStr(*PG_GETARG_NAME(1));
+	Oid			inputFuncId = PG_GETARG_OID(2);
+	Oid			outputFuncId = PG_GETARG_OID(3);
+	int16		internalLength = PG_GETARG_INT16(4);
+	char	   *probin = get_probin(fcinfo->flinfo->fn_oid);
+	bool		result;
+
+	result = create_base_type(typeNamespace, typeName, inputFuncId, outputFuncId, internalLength, probin, true);
+
+	/*
+	 * Make effects of commands visible
+	 */
+	CommandCounterIncrement();
+
+	PG_RETURN_BOOL(result);
+}
+
+/*
+ * create_base_type
+ *
+ * Create a new base type, returns true when a new base type is successfully created.
+ *
+ * if_not_exists: if true, don't fail on duplicate name, just print a notice and return false.
+ * Otherwise, fail on duplicate name.
+ */
+static bool
+create_base_type(Oid typeNamespace, char *typeName, Oid inputFuncId, Oid outputFuncId, int16 internalLength, char *funcProbin, bool if_not_exists)
+{
+	AclResult	aclresult;
+	Oid			inputOid;
+	Oid			outputOid;
+	Oid			typeOid;
+	Oid			array_oid;
+	char	   *array_type;
+	ObjectAddress address;
+	Oid			inputFuncParamType;
+	Oid			outputFuncParamType;
+	char	   *namespaceName;
+
+	/*
+	 * Even though the SQL function is locked down so only a member of
+	 * pgtle_admin can run this function, let's check and make sure there is
+	 * not a way to bypass that
+	 */
+	check_is_pgtle_admin();
+
+	if (!(internalLength > 0 || internalLength == -1))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_OBJECT_DEFINITION),
+				 errmsg("invalid type internal length %d", internalLength)));
+
+	if (internalLength > 0)
+		internalLength += VARHDRSZ;
+
+	/* Check we have creation rights in target namespace */
+	aclresult = PG_NAMESPACE_ACLCHECK(typeNamespace, GetUserId(), ACL_CREATE);
+	namespaceName = get_namespace_name(typeNamespace);
+	if (aclresult != ACLCHECK_OK)
+		aclcheck_error(aclresult, OBJECT_SCHEMA, namespaceName);
+
+	/*
+	 * Look to see if type already exists
+	 */
+#if PG_VERSION_NUM >= 120000
+	typeOid = GetSysCacheOid2(TYPENAMENSP, Anum_pg_type_oid,
+							  CStringGetDatum(typeName),
+							  ObjectIdGetDatum(typeNamespace));
+#else
+	typeOid = GetSysCacheOid2(TYPENAMENSP,
+							  CStringGetDatum(typeName),
+							  ObjectIdGetDatum(typeNamespace));
+#endif
+
+	/*
+	 * If it's not a shell, see if it's an autogenerated array type, and if so
+	 * rename it out of the way.
+	 */
+	if (OidIsValid(typeOid) && get_typisdefined(typeOid))
+	{
+		if (moveArrayTypeName(typeOid, typeName, typeNamespace))
+			typeOid = InvalidOid;
+		else if (if_not_exists)
+		{
+			ereport(NOTICE,
+					(errcode(ERRCODE_DUPLICATE_OBJECT),
+					 errmsg("type \"%s\" already exists, skipping", typeName)));
+			return false;
+		}
+		else
+			ereport(ERROR,
+					(errcode(ERRCODE_DUPLICATE_OBJECT),
+					 errmsg("type \"%s\" already exists", typeName)));
```

Comment:
This is very if/else heavy. Here's an alternative approach:
```suggestion
		if (!moveArrayTypeName(typeOid, typeName, typeNamespace))
		{
		  if (if_not_exists)
		    return false;
		    
		  ereport(NOTICE,
					(errcode(ERRCODE_DUPLICATE_OBJECT),
					 errmsg("type \"%s\" already exists", typeName)));
		}

		typeOid = InvalidOid;
```

Reply:
Done. Thanks for the suggestion.